alg/teca_regional_moisture_flux.cxx

```diff
@@ -0,0 +1,690 @@
+#include <thrust/functional.h>
+#include <thrust/transform_reduce.h>
+#include <thrust/iterator/zip_iterator.h>
+
+#endif
+
+using namespace teca_variant_array_util;
+using allocator = teca_variant_array::allocator;
+
+//#define TECA_DEBUG
+
+namespace
+{
+// CPU codes
+//
+
+namespace cpu
+{
+// **************************************************************************
+template <typename coord_t>
+void transform_coordinates(coord_t *theta, coord_t *phi, coord_t *sin_phi,
+    const coord_t *lon, const coord_t *lat, unsigned long nlon, unsigned long nlat)
+{
+    // transform the cooridnates
+    coord_t rad_deg = M_PI / 180.0;
+
+    for (unsigned long i = 0; i < nlon; ++i)
+    {
+        theta[i] = lon[i] * rad_deg;
+    }
+
+    for (unsigned long j = 0; j < nlat; ++j)
+    {
+        coord_t phi_j = (90.0 - lat[j]) * rad_deg;
+
+        phi[j] = phi_j;
+
+        sin_phi[j] = sin(phi_j);
+    }
+}
+
+
+// **************************************************************************
+template <typename coord_t, typename data_t>
+void divergence(data_t *div, const data_t *ivt_u, const data_t *ivt_v,
+    const coord_t *theta, const coord_t *phi, const coord_t *sin_phi,
+    unsigned long nlon, unsigned long nlat)
+{
+    unsigned long nlatm1 = nlat - 1;
+    unsigned long nlonm1 = nlon - 1;
+
+    // compute the divergence at every point
+    for (unsigned long j = 1; j < nlatm1; ++j)
+    {
+        for (unsigned long i = 1; i < nlonm1; ++i)
+        {
+            unsigned long q = j*nlon + i;
+
+            // div = d / d theta ( ivt_theta )
+            //     + d / d phi ( ivt_phi sin(phi) )
+
+            // forward difference. using a centered difference here resulted in
+            // a larger overall error.
+            data_t dthe = theta[i + 1] - theta[i];
+            data_t dphi = phi[j + 1] - phi[j];
+
+            data_t divt_the = ivt_u[q + 1] - ivt_u[q];
+            data_t divtsin_phi =  ivt_v[q + nlon] * sin_phi[j + 1] - ivt_v[q] * sin_phi[j];
+
+            div[q] = divt_the / dthe + divtsin_phi / dphi;
+        }
+    }
+}
+
+// **************************************************************************
+template <typename coord_t, typename data_t>
+void integrate(data_t *cell_int, const data_t *div,
+    const coord_t *theta, const coord_t *phi,
+    unsigned long nlon, unsigned long nlat)
+{
+    unsigned long nlatm1 = nlat - 1;
+    unsigned long nlonm1 = nlon - 1;
+
+    // integrate cell by cell using the trapezoid rule
+    for (unsigned long j = 1; j < nlatm1; ++j)
+    {
+        for (unsigned long i = 1; i < nlonm1; ++i)
+        {
+            unsigned long q = j*nlon + i;
+
+            // area of the cell
+            data_t dthe = theta[i + 1] - theta[i];
+            data_t dphi = phi[j + 1] - phi[j];
+            data_t dA = dthe * dphi;
+
+            // height of the surface above the cell. using trapezoid here
+            // resulted in a larger overall error.
+            cell_int[q] = div[q] * dA;
```

Comment:
shouldn't there be a multiplication by sin(phi[j]) here to account for convergence of the meridians (i.e., dA should go to 0 near the poles)?

Reply:
> shouldn't there be a multiplication by sin(phi[j]) here to account for convergence of the meridians (i.e., dA should go to 0 near the poles)?

It cancels with the factor of 1/(r_earth*sin(phi)) from the divergence. See eqn 8 in the overleaf.

Reply:
got it, thanks.  I was referring to another source for the spherical coordinate divergence, just for independence, but yes, I agree with this cancellation.